### PR TITLE
SLES16.1 Immutable host/guest installation

### PR DESCRIPTION
**Pull request by waynechen55**
* **Add** guest profile using immutable online iso:
  * ```data/virt_autotest/guest_params_xml_files/sles_16_1_immutable_64_kvm_hvm_x86_64_agama_online_iso.xml```
  * This profile is the very first one which also serves as an example, future installation can just use test suite level setting ```UNIFIED_GUEST_PRODUCT_MODES=immutable or standard or empty or omitted``` without a need to add new profile.
 
* **And** product mode option in agama jsonnet files:
  * ```data/virt_autotest/host_unattended_installation_files/agama/sle16.jsonnet```
  * ```data/virt_autotest/guest_unattended_installation_files/sles_16_plus_kvm_hvm_guest_non_encrypted_storage_x86_64.jsonnet```

* **Add** variable ```VERSION```, ```VERSION_TO_INSTALL``` and ```TRANSACTIONAL``` to be expanded in ```lib/autoyast.pm```.  And do realistic replacement in   ```data/virt_autotest/host_unattended_installation_files/agama/sle16.jsonnet``` by leveraging jsonnet language features, namely defining corresponding variables and do replacement and assignment in ```if...then...else```.

* **Show** agama profile by using ```agama config show``` in module ```tests/installation/agama_reboot.pm```, so user can know what is being used for installation explicitly, because only dynamic agama jsonnet file is uploaded in module ```tests/installation/ipxe_install.pm```.

* **Support** product mode setting for guest as well in modules
  * ```lib/guest_installation_and_configuration_base.pm```
  * ```lib/guest_installation_and_configuration_metadata.pm```
  
* **Set** guest ```guest_product_mode``` to default value ```standard``` if it is empty with SLES16.1 product onwards in ```lib/guest_installation_and_configuration_base.pm```, otherwise drop it in agama jsonnet.
  
* **Support** setting guest product mode at the test suite level by introducing setting ```UNIFIED_GUEST_PRODUCT_MODES``` in module files:
  * ```tests/virt_autotest/unified_guest_installation.pm```
  * ```lib/concurrent_guest_installations.pm```
which has the same usage, for example: ```UNIFIED_GUEST_PRODUCT_MODES='standard|immutable'```.

* **Do** preparation based on product mode in schedule file ```schedule/virt_autotest/sle16_guest_installation.yaml``` which loads
either ```virt_autotest/prepare_non_transactional_server``` or
```virt_autotest/prepare_transactional_server``` based setting ```TRANSACTIONAL```.

* **Update** ```is_sle``` to inspect ```VERSION_TO_INSTALL``` first in order to do comparison with host running not-current ```VERSION```.

* **Please** also refer to [Agama documentation](https://agama-project.github.io/docs/user/reference/profile/product) for product mode definition:
```
Some products support different installation modes. The mode can change how the product behaves, for instance, enabling an immutable OS with transactional updates.

The supported values depend on the product. For instance, SLES supports standard and immutable modes.

In an unattended installation profile, the mode key is optional for backward compatibility; if omitted, the product will use its traditional default behavior. However, it is preferred to explicitly specify the desired mode.
``` 
Explicit specification is suggested.

* **Also** review [Add baremetal immutable host/guest testing](https://gitlab.suse.de/qa-testsuites/openqa-job-group-yaml/-/merge_requests/380).

* **Verification Runs:**
  * [SLES16.1 immutable guest on immutable host](http://vh012.qa2.suse.asia/tests/652)
  * [SLES16.1 immutable guest on traditional host](http://vh012.qa2.suse.asia/tests/655)
  * [SLES16.1 traditional guest on immutable host](http://vh012.qa2.suse.asia/tests/650)
  * [SLES16.1 traditional guest on traditional host](http://vh012.qa2.suse.asia/tests/658)
  * [SLES16.0 guest on SLES16.1 traditional host](http://vh012.qa2.suse.asia/tests/632)
  * [SLES16.1 traditional guest on SLES16.0 host](http://vh012.qa2.suse.asia/tests/633)
  * [sle-16.1-Online-x86_64-Build35.2-uefi-gi-guest_sles16_0-online-on-host_developing-kvm](https://openqa.suse.de/tests/21719200)
  * [Test UNIFIED_GUEST_PRODUCT_MODES=standard|immutable](http://vh012.qa2.suse.asia/tests/636)
  * [Test UNIFIED_GUEST_PRODUCT_MODES=|immutable](http://vh012.qa2.suse.asia/tests/663)
  * [sle-16.1-Online-x86_64-Build35.2-uefi-gi-online-guest_developing-on-host_developing-kvm](https://openqa.suse.de/tests/21739762)
  * [sle-16.1-Online-x86_64-Build35.2-snapshot-gi-guest_developing-online-on-host_sles16_0-kvm](https://openqa.suse.de/tests/21739763)  
  * [sle-16.1-Online-x86_64-Build35.2-uefi-gi-guest_developing-on-host_sles16_0-kvm](https://openqa.suse.de/tests/21739764)
  * [sle-16.1-Online-x86_64-Build35.2-snapshot-gi-guest_sles16_0-online-on-host_developing-kvm](https://openqa.suse.de/tests/21739765)
  * [oraclelinux](https://openqa.suse.de/tests/21625307)
  * [SLES16.1 immutable guest on immutable host OSD installation](https://openqa.suse.de/tests/21719226)
  * [SLES16.1 immutable guest on traditional host OSD installation](https://openqa.suse.de/tests/21739767)
  * [sle-16.1-Online-x86_64-Build35.2-snapshot-gi-guest_developing-on-host_sles15sp7-kvm](https://openqa.suse.de/tests/21739768)
  * [sle-16.1-Full-ppc64le-Build35.2-sles_default_immutable](https://openqa.suse.de/tests/21720551)
  * [sle-16.1-Full-x86_64-Build35.2-sles_offline_oemdrv_unattended](https://openqa.suse.de/tests/21720552)
  * [sle-16.1-Full-ppc64le-Build35.2-sles_sap_default](https://openqa.suse.de/tests/21720554)
  * [sle-16.1-Online-aarch64-Build35.2-sles_autoyast_supported](https://openqa.suse.de/tests/21739769)
  * [sle-16.1-Online-ppc64le-Build35.2-sles_autoyast_unsupported](https://openqa.suse.de/tests/21739770)
  * [sle-16.1-Online-x86_64-Build35.2-sles_autoyast_unsupported](https://openqa.suse.de/tests/21739771)
  * [sle-16.1-Online-x86_64-Build35.2-sles_autoyast_supported](https://openqa.suse.de/tests/21739772)
  * [ sle-16.1-Online-aarch64-Build35.2-sles_default_immutable_unattended](https://openqa.suse.de/tests/21739774)
  * [sle-16.1-Online-ppc64le-Build35.2-sles_default_immutable_unattended](https://openqa.suse.de/tests/21739775)
  * [sle-16.1-Online-s390x-Build35.2-sles_default_immutable_unattended](https://openqa.suse.de/tests/21739776)
  * [sle-16.1-Online-x86_64-Build35.2-sles_default_immutable_unattended](https://openqsle-16.1-Full-ppc64le-Build35.2-agama_create_hdd_textmode_qesec_hmca.suse.de/tests/21739777)
  * [sle-16.1-Online-aarch64-Build35.2-sles_gnome_unattended](https://openqa.suse.de/tests/21739778)
  * [sle-16.1-Online-s390x-Build35.2-sles_gnome_unattended](https://openqa.suse.de/tests/21739779)
  * [sle-16.1-Online-x86_64-Build35.2-sles_gnome_unattended](https://openqa.suse.de/tests/21739780)
  * [sle-16.1-Transactional-Base-qcow-x86_64-Build5.5-combustion](https://openqa.suse.de/tests/21739782)
  * [sle-16.1-Transactional-Base-ppc-4096-ppc64le-Build5.5-cloud-init](https://openqa.suse.de/tests/21739783)
  * [sle-16.1-Online-x86_64-Build35.2-blktests](https://openqa.suse.de/tests/21739784)
  * [sle-16.1-Online-s390x-Build35.2-install_ltp+sle+Online](https://openqa.suse.de/tests/21739785)
  * [sle-16.1-Online-ppc64le-Build35.2-kdump](https://openqa.suse.de/tests/21739788)
  * [sle-16.1-Online-aarch64-Build35.2-ltp_power_management_tests](https://openqa.suse.de/tests/21739789)
  * [sle-16.1-Online-ppc64le-Build35.2-nfs_cthon04-nfs4](https://openqa.suse.de/tests/21739790)
  * [sle-16.1-Online-x86_64-Build35.2-qa_kernel_kexec](https://openqa.suse.de/tests/21739791)
  * [sle-16.1-Online-Kernel-64kb-Baremetal-aarch64-Build35.2-ltp_syscalls](https://openqa.suse.de/tests/21739792)
  * [sle-16.1-Online-Kernel-Baremetal-x86_64-Build35.2-install_ltp+sle+Online-Kernel-Baremetal](https://openqa.suse.de/tests/21739793)
  * [ sle-16.1-Online-Kernel-Baremetal-ppc64le-Build35.2-ltp_openposix](https://openqa.suse.de/tests/21739794)
  * [sle-16.1-Online-Kernel-Base-x86_64-Build35.2-ltp_syscalls](https://openqa.suse.de/tests/21739795)
  * [sle-16.1-Online-Kernel-RT-x86_64-Build35.2-ltp_dio](https://openqa.suse.de/tests/21739796)
  * [sle-16.1-Online-Kernel-RT-Baremetal-x86_64-Build35.2-kdump](https://openqa.suse.de/tests/21739797)
  * [sle-16.1-Online-x86_64-Build35.2-wayland-desktopapps-firefox](https://openqa.suse.de/tests/21739798)
  * [sle-16.1-Online-x86_64-Build35.2-sles_migration_15sp4_ha](https://openqa.suse.de/tests/21740084)
  * [sle-16.1-Online-x86_64-Build35.2-sles_migration_15sp5_textmode](https://openqa.suse.de/tests/21740085)
  * [sle-16.1-Online-x86_64-Build35.2-sles_migration_15sp5_textmode_all_modules_and_extensions](https://openqa.suse.de/tests/21740086)
  * [sle-16.1-Online-x86_64-Build35.2-sles_migration_15sp7_ha](https://openqa.suse.de/tests/21740089)
  * [sle-16.1-Online-x86_64-Build35.2-sles_migration_15sp7_textmode_all_modules_and_extensions](https://openqa.suse.de/tests/21740090)
  * [sle-16.1-Online-x86_64-Build35.2-sles_migration_160](https://openqa.suse.de/tests/21740091)
  * [sle-16.1-Online-x86_64-Build35.2-sles_nvme_unattended](https://openqa.suse.de/tests/21740092)
  * [sle-16.1-Online-x86_64-Build35.2-sles_root_filesystem_xfs_unattended](https://openqa.suse.de/tests/21740093)
  * [sle-16.1-Online-x86_64-Build35.2-sles_sap_default_unattended](https://openqa.suse.de/tests/21740094)
  * [sle-16.1-Online-x86_64-Build35.2-sles_sap_default_immutable_unattended](https://openqa.suse.de/tests/21740095)
  * [sle-16.1-Full-ppc64le-Build35.2-sles_default_immutable](https://openqa.suse.de/tests/21740097)
  * [sle-16.1-Full-ppc64le-Build35.2-sles_sap_default_unattended](https://openqa.suse.de/tests/21740103)
  * [sle-16.1-Full-x86_64-Build35.2-sles_offline_oemdrv_unattended](https://openqa.suse.de/tests/21740104)
  * [sle-16.1-Online-aarch64-Build35.2-sles_autoyast_supported](https://openqa.suse.de/tests/21740105)
  * [sle-16.1-Online-ppc64le-Build35.2-sles_autoyast_unsupported](https://openqa.suse.de/tests/21740106)
  * [sle-16.1-Online-s390x-Build35.2-sles_default_headless](https://openqa.suse.de/tests/21740109)
  * [sle-16.1-Online-s390x-Build35.2-sles_default_standard](https://openqa.suse.de/tests/21740110)
  * [sle-16.1-Online-ppc64le-Build35.2-sles_default_unattended](https://openqa.suse.de/tests/21740111)
  * [sle-16.1-Online-ppc64le-Build35.2-sles_extended_unattended](https://openqa.suse.de/tests/21740113)
  * [sle-16.1-Online-s390x-Build35.2-sles_gnome_unattended](https://openqa.suse.de/tests/21740114)
  * [sle-16.1-Online-x86_64-Build35.2-sles_iscsi_ibft_lvm_encrypted_unattended](https://openqa.suse.de/tests/21740115)
  * [sle-16.1-Online-s390x-Build35.2-sles_lvm_encrypted_unattended](https://openqa.suse.de/tests/21740116)
  * [sle-16.1-Online-ppc64le-Build35.2-sles_lvm_unattended](https://openqa.suse.de/tests/21740117)
  * [sle-16.1-Online-aarch64-Build35.2-sles_migration_15sp5_textmode_all_modules_and_extensions](https://openqa.suse.de/tests/21740118)
  * [sle-16.1-Online-ppc64le-Build35.2-sles_migration_15sp6_textmode](https://openqa.suse.de/tests/21740119)
  * [sle-16.1-Online-s390x-Build35.2-sles_migration_15sp6_textmode_all_modules_and_extensions](https://openqa.suse.de/tests/21740120)
  * [sle-16.1-Online-s390x-Build35.2-sles_migration_160](https://openqa.suse.de/tests/21740121)
  * [sle-16.1-Online-x86_64-Build35.2-sles_migration_15sp7_textmode](https://openqa.suse.de/tests/21740122)
  * [sle-16.1-Online-ppc64le-Build35.2-sles_sap_migration_160](https://openqa.suse.de/tests/21740124)
  * [sle-16.1-Online-aarch64-Build35.2-sles_tpm_fde_unattended](https://openqa.suse.de/tests/21740128)
  * [sle-16.1-Online-aarch64-Build35.2-aardvark_testsuite](https://openqa.suse.de/tests/21740283)
  * [sle-16.1-Online-ppc64le-Build35.2-containerd_testsuite](https://openqa.suse.de/tests/21740284)
  * [sle-16.1-Online-s390x-Build35.2-docker_rootless_testsuite](https://openqa.suse.de/tests/21740285)
  * [sle-16.1-Online-x86_64-Build35.2-docker_testsuite](https://openqa.suse.de/tests/21740286)
  * [sle-16.1-Online-x86_64-Build35.2-multi_runtime](https://openqa.suse.de/tests/21740290)
  * [sle-16.1-Online-x86_64-Build35.2-netavark_testsuite](https://openqa.suse.de/tests/21740291)
  * [sle-16.1-Online-s390x-Build35.2-podman_tests](https://openqa.suse.de/tests/21740292)
  * [sle-16.1-Online-aarch64-Build35.2-skopeo_testsuite](https://openqa.suse.de/tests/21740293)
  * [sle-16.1-Online-ppc64le-Build35.2-runc_testsuite](https://openqa.suse.de/tests/21740294)
  * [sle-16.1-Online-aarch64-Build35.2-aardvark_testsuite](https://openqa.suse.de/tests/21741651)
  * [sle-16.1-Online-ppc64le-Build35.2-containerd_testsuite](https://openqa.suse.de/tests/21741652)
  * [sle-16.1-Online-s390x-Build35.2-docker_rootless_testsuite](https://openqa.suse.de/tests/21741653)
  * [sle-16.1-Online-x86_64-Build35.2-multi_runtime](https://openqa.suse.de/tests/21741654)
  * [sle-16.1-Online-x86_64-Build35.2-podman_e2e](https://openqa.suse.de/tests/21741655)
  * [sle-16.1-Online-s390x-Build35.2-podman_testsuite](https://openqa.suse.de/tests/21741656)
  * [sle-16.1-Online-ppc64le-Build35.2-skopeo_testsuite](https://openqa.suse.de/tests/21741657)
  * [sle-16.1-Online-aarch64-Build35.2-skopeo_testsuite](https://openqa.suse.de/tests/21741658)
  * [sle-16.1-Full-x86_64-Build35.2-sles4sap_offline_dvd_textmode](https://openqa.suse.de/tests/21741659)
  * [sle-16.1-Full-ppc64le-Build35.2-sles4sap_online_dvd_textmode](https://openqa.suse.de/tests/21741660)
  * [sle-16.1-Online-x86_64-Build35.2-create_hdd_sles4sap](https://openqa.suse.de/tests/21741661)
  * [sle-16.1-Online-x86_64-Build35.2-patterns_sles4sap](https://openqa.suse.de/tests/21741662)
  * [sle-16.1-Online-ppc64le-Build35.2-sles4sap_hmc_textmode_netweaver](https://openqa.suse.de/tests/21741663)
  * [sle-16.1-Online-x86_64-Build35.2-sles4sap_scc_textmode_netweaver](https://openqa.suse.de/tests/21741665)
  * [sle-16.1-Minimal-VM-4096-sap-ppc64le-Build13.24-minimalvm-main-combustion](https://openqa.suse.de/tests/21741716)
  * [sle-16.1-Minimal-VM-512-for-kvm-and-xen-ppc64le-Build13.24-minimalvm-cloud-init](https://openqa.suse.de/tests/21741720)
  * [sle-16.1-Minimal-VM-512-sap-ppc64le-Build13.24-minimalvm-main](https://openqa.suse.de/tests/21741720)
  * [sle-16.1-Minimal-VM-Cloud-x86_64-Build13.24-minimalvm-cloud-init](https://openqa.suse.de/tests/21741721)
  * [sle-16.1-Minimal-VM-Cloud-x86_64-Build13.24-minimalvm-cloud-init](https://openqa.suse.de/tests/21741722)
  * [sle-16.1-Minimal-VM-Cloud-sap-x86_64-Build13.24-minimalvm-wizard](https://openqa.suse.de/tests/21741723)
  * [sle-16.1-Transactional-Base-aarch64-Build5.55-ignition](https://openqa.suse.de/tests/21741725)
  * [sle-16.1-Transactional-Base-RT-x86_64-Build5.55-combustion](https://openqa.suse.de/tests/21741727)
  * [sle-16.1-Transactional-Base-RT-x86_64-Build5.55-ignition](https://openqa.suse.de/tests/21741728)
  * [sle-16.1-Transactional-Base-SelfInstall-x86_64-Build5.55-combustion](https://openqa.suse.de/tests/21741730)
  * [sle-16.1-Transactional-Base-qcow-aarch64-Build5.55-combustion@](https://openqa.suse.de/tests/21741732)
  * [sle-16.1-Transactional-Base-qcow-x86_64-Build5.55-ignition](https://openqa.suse.de/tests/21741733)
  * [sle-16.1-Online-x86_64-Build35.2-blktests](https://openqa.suse.de/tests/21741734)
  * [sle-16.1-Online-s390x-Build35.2-install_ltp+sle+Online](https://openqa.suse.de/tests/21741735)
  * [sle-16.1-Online-aarch64-Build35.2-ltp_aio_stress](https://openqa.suse.de/tests/21741736)
  * [sle-16.1-Online-ppc64le-Build35.2-ltp_irq](https://openqa.suse.de/tests/21741737)
  * [sle-16.1-Online-aarch64-Build35.2-nfs_cthon04-nfs4](https://openqa.suse.de/tests/21741740)
  * [sle-16.1-Online-ppc64le-Build35.2-nfs_pynfs_nfs41_all](https://openqa.suse.de/tests/21741741)
  * [sle-16.1-Online-x86_64-Build35.2-qa_kernel_multipath_supportserver](https://openqa.suse.de/tests/21741742)
  * [sle-16.1-Online-x86_64-Build35.2-qa_kernel_multipath](https://openqa.suse.de/tests/21741743)
  * [sle-16.1-Online-Kernel-Baremetal-ppc64le-Build35.2-fadump](https://openqa.suse.de/tests/21741749)
  * [sle-16.1-Online-Kernel-Baremetal-x86_64-Build35.2-install_ltp+sle+Online-Kernel-Baremetal](https://openqa.suse.de/tests/21741750)
  * [sle-16.1-Online-Kernel-Base-x86_64-Build35.2-ltp_syscalls](https://openqa.suse.de/tests/21741754)
  * [sle-16.1-Online-Kernel-RT-x86_64-Build35.2-ltp_uevent_numa](https://openqa.suse.de/tests/21741755)
  * [sle-16.1-Online-Kernel-RT-Baremetal-x86_64-Build35.2-install_ltp+sle+Online-Kernel-RT-Baremetal](https://openqa.suse.de/tests/21741756)
  * [sle-16.1-Full-ppc64le-Build35.2-agama_create_hdd_textmode_qesec_hmc](https://openqa.suse.de/tests/21741766)
  * [sle-16.1-Online-ppc64le-Build35.2-agama_create_hdd_textmode_qesec](https://openqa.suse.de/tests/21741771)
  * [sle-16.1-Online-s390x-Build35.2-agama_create_hdd_textmode_qesec](https://openqa.suse.de/tests/21741772)
  * [sle-16.1-Online-ppc64le-Build35.2-sles_rmt_unattended](https://openqa.suse.de/tests/21741777)
  * [sle-16.1-Online-aarch64-Build35.2-unlock_luks2_volume_with_tpm2](https://openqa.suse.de/tests/21741779)
  * [sle-16.1-Online-ppc64le-Build35.2-xfstests_btrfs-generic-501-600](https://openqa.suse.de/tests/21744149)
  * [sle-16.1-Online-s390x-Build35.2-create_hdd_ha_textmode](https://openqa.suse.de/tests/21744153)
  * [sle-16.1-Online-ppc64le-Build35.2-rear_generate_backup_textmode](https://openqa.suse.de/tests/21744154)

### CODE REVIEW

**Comment by Julie-CAO** [comment]
Adding a setting to the host Agama profile and the test module scheduler looks fine to me.

**Comment by waynechen55** [comment]
> * for guest, use UNIFIED_GUEST_PRODUCT_MODES, and it can be used freely regardless of whether sle16.1 host is standard or immutable, right?

Yes.


> * AGAMA_PRODUCT_MODE and TRANSACTIONAL: used to control host, right? They are supposed to be used with constraints with each other, although they are two separate settings? That is "TRANSACTIONAL=1" should work with AGAMA_PRODUCT_MODE=immutable, right? If so, shall we add some checks for them to avoid mismatching settings?

- Yes, they are for host control.
- ```TRANSACTIONAL=0 or 1``` must be set if test suite uses schedule file ```schedule/virt_autotest/sle16_guest_installation.yaml```, otherwise there will be no preparation module to be loaded. Otherwise, standard host does not have to set ```TRANSACTIONAL=0```.
- Basically there is no constraint imposed from either ```TRANSACTIONAL``` or ```AGAMA_PRODUCT_MODE``` in terms of settings alone, but if host is immutable mode installed from agama iso, then you have to specify ```AGAMA_PRODUCT_MODE=immutable``` which leads to ```TRANSACTIONAL=1``` has to be set for test suite because subroutine ```is_transactional``` relies on this setting. So this results from the test scenario which imposes the logic.
- There are some common combinations, for example:
  - ```TRANSACTIONAL=0``` alone
  - ```TRANSACTIONAL=0 and  AGAMA_PRODUCT_MODE=standard```
  - ```TRANSACTIONAL=1 and AGAMA_PRODUCT_MODE=immutable```
  All the other combinations do not have meaningful effect.
- Mismatch settings do not have meaningful effect, because:
  - ```TRANSACTIONAL=1``` on standard host will try to use ```transactional-update shell``` which does not exist at all
  - ```TRANSCATIONAL=0``` on immutable host will try to modify files can not be modified directly which leads to definite failure.
  - These mismatched settings leads to wrong preparation module to be loaded the result of which is just only failure. This is determined by the schedule file ```schedule/virt_autotest/sle16_guest_installation.yaml```, so there is no verification needed for them.

**Comment by alice-suse** [comment]
> > > > Yes, exactly. Mismatched settings lead to unreasonable result or failure at latter modules. So I gave the suggestion to check them at initial step -- if mismatched, die earlier(eg before the ipxe starts).
> > > 
> > > 
> > > User should not use `TRANSACTIONAL=1 AGAMA_PRODUCT_MODE=standard` or `TRANSACTIONAL=0 AGAMA_PRODUCT_MODE=immutable`. If you want to have this scenario, please help provide reasoning.
> > 
> > 
> > We never needs such scenario, of course. You seems not get the point yet. Let me try to rephrase...
> > Your current solution give the "SHOULD NOT" requirements to users, while for programming, for such "inputs", we can add the check in code and just let it fail directly when not matched "inputs" given, right? This is not big deal, just a minor suggestion, since not every user is familiar with what are correct combinations and what are not, as you do.
> 
> This is not requirement to anyone. This is definition. I do not think there is necessity to do so. And I also have a minor suggestion, do not try to treat any code change as a "requirement" or "burden" to anyone. This complicates things. I can help add new settings to doc variables.md which is the place to handle such issue.

To correctly use product and the corresponding automation test of it, of course there are requirements. Nothing negative ;-) . Adding to variables.md is an useful way to decrease wrong usage. Let you decide.

File: lib/autoyast.pm [hunk 1]
```diff
@@ -729,8 +729,12 @@ sub expand_variables {
...
             set_var('WORKER_IP', inet_ntoa(inet_aton(get_var 'WORKER_HOSTNAME')));
         }
-        # Skip if value is not defined
-        next unless my ($value) = get_var($var);
+        # Skip if value is not defined. For AGAMA_PRODUCT_MODE, use 'standard' for sle16.1+ or drop it otherwise.
+        my ($value) = get_var($var);
+        unless ($value) {
+            (is_sle('>=16.1') and $var eq 'AGAMA_PRODUCT_MODE') ? $profile =~ s/\{\{$var\}\}/standard/g : $profile =~ s/.*\{\{$var\}\}.*\n?//g;
+            next;
+        }
```
[the remark this answers is not in the document]
Comment (waynechen55) [reply lib/autoyast.pm]: I am doing some verification picked from other job groups which would be enough to cover all cases.

**Review by jknphy — APPROVED** [review]

**Comment by waynechen55** [comment]
Finished all verification runs. Please refer to the pull request description.

**Comment by alice-suse** [comment]
LGTM

**Comment by alice-suse** [comment]
Thanks to @jknphy for the suggestions related with agama!
